File: programme/matrix-multiplication/matrix_multiplication.py
'''
PROBLEM STATEMENT
Write a program to print multiplication of 2 matrices.

First matrix
1  1  1
2  2  2 
3  3  3 

Second matrix
1  1  1
2  2  2 
3  3  3 

Product matrix
6  6  6
12 12 12
18 18 18
'''
# SOLUTION --->

# first matrix (3x3)
A = [[1, 1, 1],
     [2, 2, 2],
     [3, 3, 3]]

# second matrix
B = [[1, 1, 1],
     [2, 2, 2],
     [3, 3, 3]]

# output form
result = [[0, 0, 0],
          [0, 0, 0],
          [0, 0, 0]]

# iterating by row of A
for i in range(len(A)):

    # iterating by column by B
    for j in range(len(B[0])):

        # iterating by rows of B
        for k in range(len(B)):
            result[i][j] += A[i][k] * B[k][j]

for r in result:
    print(r)

File: programme/matrix-multiplication/matrix_multiplication.go
// Write a programme to print multiplication of 2 matrices
package main

import (
	"fmt"
)

func displayMatrix(m [][]int) {
	// show the matrix content in tabular form
	var row int = len(m)
	var col int = len(m[0])
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			fmt.Printf("%d ", m[i][j])
		}
		fmt.Printf("\n")
	}
	fmt.Printf("\n")
}

func matrixMultiplication(a, b [][]int) {
	// multiplies two matrices and show the resultant matrix
	var row int = len(a)
	var col int = len(a[0])
	var result = make([][]int, row)
	for i := 0; i < row; i++ {
		result[i] = make([]int, col)
	}
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			result[i][j] = 0
			for k := 0; k < row; k++ {
				result[i][j] += a[i][k] * b[k][j]
			}
		}
	}
	fmt.Printf("First Matrix: \n")
	displayMatrix(a)
	fmt.Printf("Second Matrix: \n")
	displayMatrix(b)
	fmt.Println("Product Matrix: ")
	displayMatrix(result)
}

func main() {
	var first_matrix = [][]int{
		{1, 1, 1},
		{2, 2, 2},
		{3, 3, 3},
	}
	var second_matrix = [][]int{
		{1, 1, 1},
		{2, 2, 2},
		{3, 3, 3},
	}

	matrixMultiplication(first_matrix, second_matrix)
}
